### DISCUSSION TITLE
Quest  - Add a config for a Configurable Patient List view that can be adapted for either EIR, RDT or ANC apps

### DISCUSSION THREAD

**Discussion by f-odhiambo** [opensrp/fhircore#1137] [Ideas]
**Describe the issue to be researched**
Add a config that can be able to reuse the patient list view properties for each health vertical 

**Describe the goal of the research**
 Try to figure out how we can have a structured config to enable and disable text views and labels on  the Patient list view 

**Describe the methodology**

1. The purpose of this RnD is to figure out  how to add and structure a config that can enable and disable properties on the UI, based on the credential of the user and the selected health verifcal 
2. The config should be able to accommodate all use cases delivered so far such as ANC, EIR and G6PD RDT app. That means if I load the ANC app I can be able to view on the Patient List view the Due and Upcoming services, while if I load the G6PD app these are disabled or if I load the Covax app I can be able to view all Due patient with overdue vaccination status

**Comment by f-odhiambo** [comment]
This ticket is to facilitate some RnD ahead of our upcoming dicisission

**Reply by ellykits** [reply]
Some questions @f-odhiambo

1. Do we want to use one view to present data for the different health verticals (ANC/EIR/G6PD etc)
2. If above is yes how should we go about filtering data to display in the views?
3. Will we have a situation where we want to display register data for multiple health verticals for instance a supervisor user wants to view both ANC and EIR registered patients
4. Where should we declare the different supported health verticals? Should they be keycloak user roles? Or do we have to hardcode them in the composition resource/application configuration?
5. In line with above number 4 do we want to apply this configuration at the data level or application level? 

CC @dubdabasoduba @pld

**Comment by ellykits** [comment]
**Applying configuration at the Data Level**

Sample ANC Patient resource

```json
{
  "resourceType": "Patient",
  "id": "charity-otala-1",
  "meta": {
    "versionId": "21",
    "lastUpdated": "2022-03-05T20:52:48.295+00:00",
    "source": "#aa5851bce3d30673",
    "profile": [
      "http://fhir.org/guides/who/anc-cds/StructureDefinition/ancpatient"
    ],
    "tag": [
      {
        "system": "https://www.snomed.org",
        "code": "35359004",
        "display": "Family"
      },
      {
        "system": "http://loinc.org",
        "code": "LA15173-0",
        "display": "Pregnant"
      }
    ]
  }
}
```
- We can use this tag to flag ANC registered patients and do the same for the other health verticals
```json 
 {
        "system": "http://loinc.org",
        "code": "LA15173-0",
        "display": "Pregnant"
  }
```
Then load the correct patient list view row configs; depending on the specified tag.

**Comment by f-odhiambo** [comment]
**Notes Mar 10th, 2022** - Meeting @ellykits @dubdabasoduba 

1. Need to agree on all the current health verticals supported 
    - ANC, EIR, CHW
2. Need to write some config to identify the UI element for all the different health verticals, loaded at runtime, away from the composition resource
3. Composition resource will be used to bundle all the required resources for building an app on a health vertical 
4. How will  Composition resource compliment the RBAC for enabling and disabling the different parts of the views, and what the user can be able to do or not do?
5. How do we ensure that we only get CarePlans referenced by the CareTeam of the Practioner and filter out any persons who do not belong to that vertical e.g. ANC ?

**Comment by ellykits** [comment]
Recommended model for health module configuration.

```json
{
  "defaultFeatures": [
    {
      "name": "PatientManagement",
      "active": true,
      "settings": {}
    }
  ],
  "healthModules": [
    {
      "healthModule": "RDT",
      "active": false
    },
    {
      "healthModule": "CHW",
      "active": true,
      "features": [
        {
          "name": "Referral",
          "active": true,
          "settings": {}
        }
      ]
    },
    {
      "healthModule": "ANC",
      "active": true,
      "features": [
        {
          "name": "InAppReporting",
          "active": true,
          "settings": {
            "dailyTallies": false
          }
        }
      ]
    },
    {
      "healthModule": "PNC",
      "active": false
    },
    {
      "healthModule": "EIR",
      "active": false
    }
  ]
}
```
**Note**
- Where `settings` is a key-value pair map of configs used to override the health module feature configs.
- `defaultFeatures` settings can be overridden via the health module features
- When a health module is activated it's workflow will be used on the features.
- Features will be independent; there can only be one feature type in the entire application. E.g. `PatientManagement`
- The health module configuration will be attached to the `Composition` resource
- This configuration is independent of `ViewConfiguration`

**Comment by dubdabasoduba** [comment]
@ellykits This is how I am looking at this. 

**OpenSRP App Types**
---
- HF
- CHW

**HF Features**
---
- ANC 
   - Add patient
   - Record contacts (Scheduling & tasking)
   - Pregnancy outcome (Close ANC) 
- PNC
   - Add patient
   - PNC folow up (Scheduling & tasking)
   - Close PNC (Child is 28 days old)
- Family Planning
   - Add patient
   - Depending on the FP method (Scheduling & Tasking)
   - Close family planning record
- Child 
   - Add child
   - Child immunization, Vitamins & deworming (Scheduling & Taksing) 
- In-app reporting

**CHW Features**
---
- Family Module
   - Add family 
   - Add a family member
   - Family profile
   - Familly member profile
   - Remove family
   - Remove a family member
   - All Family & family member related schedules & Tasks
- In-app reporting.  
- Mapping

The one thing to remember is for CHW the entry point for patients is through a family. 

Not sure I am making very much sense. @rowo @HenryRae @pld  Does this make sense. We are trying to categorize the different functionalities into features in order to write a configurable UI for each.

cc: @f-odhiambo

**Reply by ellykits** [reply]
Thanks for this @dubdabasoduba

**Reply by ellykits** [reply]
I will update the configuration JSON schema and share

**Reply by ellykits** [reply]
```json
{
  "appId": "uniqueAppId",
  "appFeatures": [
    {
      "feature": "PatientManagement",
      "active": true,
      "settings": {},
      "target": "CHW",
      "healthModule": "ANC",
      "useCases": [
        "PATIENT_REGISTRATION",
        "ANC_VISITS",
        "PREGNANCY_OUTCOME"
      ]
    },
    {
      "feature": "PatientManagement",
      "active": true,
      "settings": {},
      "target": "CHW",
      "healthModule": "PNC",
      "useCases": [
        "PATIENT_REGISTRATION",
        "PNC_VISITS",
        "CLOSE_PNC_RECORD"
      ]
    },
     {
      "feature": "PatientManagement",
      "active": true,
      "settings": {},
      "target": "CHW",
      "healthModule": "FAMILY_PLANNING",
      "useCases": [
        "PATIENT_REGISTRATION",
        "FAMILY_PLANNING_VISITS",
        "CLOSE_FAMILY_PLANNING_RECORD"
      ]
    },
    {
      "feature": "PatientManagement",
      "active": true,
      "settings": {},
      "target": "HF",
      "healthModule": "CHILD",
      "useCases": [
        "PATIENT_REGISTRATION",
        "CHILD_IMMUNIZATION"
      ]
    },
    {
      "feature": "HouseHoldManagement",
      "active": true,
      "settings": {},
      "target": "CHW",
      "healthModule": "FAMILY",
      "useCases": [
        "HOUSEHOLD_REGISTRATION",
        "REMOVE_HOUSEHOLD",
        "HOUSEHOLD_VISITS",
        "REMOVE_HOUSEHOLD_MEMBER"
      ]
    }
  ]
}
```
